traits/ctraits.c

```diff
@@ -2970,22 +2970,71 @@ trait_traverse(trait_object *trait, visitproc visit, void *arg)
     return 0;
 }
 
+/*-----------------------------------------------------------------------------
+|  Identify double underscore names like "__qualname__" and "__package__".
+|
+|  Returns 1 if name starts and ends with "__", else 0. Returns -1 with
+|  an exception set on failure.
++----------------------------------------------------------------------------*/
+
+static int
+is_dunder_name(PyObject *name)
+{
+    Py_ssize_t name_length;
+    int kind;
+    void *data;
+
+    if (PyUnicode_READY(name) < 0) {
+        return -1;
+    }
+
+    name_length = PyUnicode_GET_LENGTH(name);
+    kind = PyUnicode_KIND(name);
+    data = PyUnicode_DATA(name);
+    return (
+        (name_length >= 2)
+        && (PyUnicode_READ(kind, data, 0) == '_')
+        && (PyUnicode_READ(kind, data, 1) == '_')
+        && (PyUnicode_READ(kind, data, name_length - 2) == '_')
+        && (PyUnicode_READ(kind, data, name_length - 1) == '_')
+    );
+}
+
 /*-----------------------------------------------------------------------------
 |  Handles the 'getattr' operation on a 'CTrait' instance:
 +----------------------------------------------------------------------------*/
 
 static PyObject *
 trait_getattro(trait_object *obj, PyObject *name)
 {
-    PyObject *value = PyObject_GenericGetAttr((PyObject *)obj, name);
+    PyObject *value;
+    int is_dunder;
+
+    value = PyObject_GenericGetAttr((PyObject *)obj, name);
     if (value != NULL || !PyErr_ExceptionMatches(PyExc_AttributeError)) {
         return value;
     }
-
     PyErr_Clear();
 
-    Py_INCREF(Py_None);
-    return Py_None;
+    /* The attribute lookup failed. For backwards compatibility, we return
+       `None` in this case, *except* for attribute names that start and end
+       with a double underscore. Those attribute names may have special meaning
+       to the interpreter, and `None` may not be a valid value, so we allow the
+       exception to propagate. */
+
+    is_dunder = is_dunder_name(name);
+    if (is_dunder < 0) {
+        return NULL;
+    }
+    else if (is_dunder) {
+        /* Retry the attribute lookup so that we get the correct exception
+           message. */
+        return PyObject_GenericGetAttr((PyObject *)obj, name);
```

Comment:
IIUC, we're having to retry in this clause instead of simply removing the `PyErr_Clear` outside of the conditional statements because the other two clauses still need/want to use `PyErr_Clear`. Is that correct?

Reply:
Hmm. This is really just laziness. What I want to do here is re-raise the _original_ `AttributeError`. I don't want to create my own exception message, since the one from the interpreter is likely to be better (and may even change with more recent versions of Python - there's a lot of work on error messages right now).

I can't simply remove the `PyErr_Clear` because `is_dunder_name` might set an exception, and it's an error to set another exception if there's already an exception set (or at least, I can't find any evidence that it's okay to do so, and there's CPython core code that goes out of its way to avoid this).

The right way to handle this without the retry is to do a `PyErr_Fetch` / `PyErr_Restore` pairing. I'll aim to update the PR to do that.

Reply:
> and it's an error to set another exception if there's already an exception set (or at least, I can't find any evidence that it's okay to do so, and there's CPython core code that goes out of its way to avoid this).

Hmm, actually, looking into the source, this may be okay. Investigating ...

Reply:
From the source, all the exception setting mechanisms end up at `_PyErr_SetObject`, which, if an exception is already set, applies implicit exception chaining. I think that's good enough for this case. I've reworked the code accordingly.

Reply:
> From the source

I didn' mean to send you down a rabbit hole 😅